cpp/src/IceBridge/IceBridge.cpp

```diff
@@ -248,263 +173,237 @@ class BridgeService : public Service
+                           function<void(bool, const pair<const Byte*, const Byte*>&)> response,
+                           function<void(exception_ptr)> error,
                            const Current& current)
 {
     //
     // We've received an invocation, either from the client via the incoming connection, or from
     // the server via the outgoing (bidirectional) connection. The current.con member tells us
     // the connection over which the request arrived.
     //
-    IceUtil::Mutex::Lock lock(_lock);
+    lock_guard<mutex> lg(_lock);
     if(_exception)
     {
-        cb->ice_exception(*_exception.get());
+        error(_exception);
     }
     else if(!_outgoing)
     {
         //
         // Queue the invocation until the outgoing connection is established.
         //
         assert(current.con == _incoming);
-        _queue.push_back(new QueuedInvocation(cb, paramData, current));
+        _queue.emplace_back(inParams, move(response), move(error), current);
     }
     else
     {
-        send(current.con == _incoming ? _outgoing : _incoming, cb, paramData, current);
+        send(current.con == _incoming ? _outgoing : _incoming, inParams, response, error, current);
     }
 }
 
 void
-BridgeConnection::send(const ConnectionPtr& dest,
-                       const AMD_Object_ice_invokePtr& cb,
-                       const pair<const Byte*, const Byte*>& paramData,
+BridgeConnection::send(const shared_ptr<Connection>& dest,
+                       pair<const Byte*, const Byte*> inParams,
+                       function<void(bool, const pair<const Byte*, const Byte*>&)> response,
+                       function<void(exception_ptr)> error,
                        const Current& current)
 {
     try
     {
         //
         // Create a proxy having the same identity as the request.
         //
-        ObjectPrx prx = dest->createProxy(current.id);
+        auto prx = dest->createProxy(current.id);
 
-        //
-        // Examine the request to determine whether it should be forwarded as a oneway or a twoway.
-        //
-        Callback_Object_ice_invokePtr d;
         if(!current.requestId)
         {
+            // Oneway request
             if(prx->ice_isTwoway())
             {
                 prx = prx->ice_oneway();
             }
-            d = newCallback_Object_ice_invoke(new Invocation(cb), &Invocation::exception, &Invocation::sent);
+            prx->ice_invokeAsync(current.operation, current.mode, inParams, nullptr, error,
+                                 [response = move(response)](bool){ response(true, make_pair(nullptr, nullptr)); },
+                                 current.ctx);
         }
         else
         {
-            d = newCallback_Object_ice_invoke(new Invocation(cb), &Invocation::success, &Invocation::exception);
+            // Twoway request
+            prx->ice_invokeAsync(current.operation, current.mode, inParams, response, error, nullptr, current.ctx);
```

Comment:
Shouldn't we also use `move` here for `response` and `error`?

Reply:
Yes, fixed.